optuna/samplers/_nsgaiii.py

```diff
@@ -0,0 +1,637 @@
+                    target_population_size,
+                    elite_population_num,
+                    population,
+                    closest_reference_points,
+                    distance_reference_points,
+                    self._rng,
+                )
+                elite_population.extend(additional_elite_population)
+                break
+        return elite_population
+
+    def _fast_non_dominated_sort(
+        self,
+        population: list[FrozenTrial],
+        directions: list[optuna.study.StudyDirection],
+    ) -> list[list[FrozenTrial]]:
+        if self._constraints_func is not None:
+            for _trial in population:
+                _constraints = _trial.system_attrs.get(_CONSTRAINTS_KEY)
+                if _constraints is None:
+                    continue
+                if np.any(np.isnan(np.array(_constraints))):
+                    raise ValueError("NaN is not acceptable as constraint value.")
+
+        dominated_count: defaultdict[int, int] = defaultdict(int)
+        dominates_list = defaultdict(list)
+
+        dominates = _dominates if self._constraints_func is None else _constrained_dominates
+
+        for p, q in itertools.combinations(population, 2):
+            if dominates(p, q, directions):
+                dominates_list[p.number].append(q.number)
+                dominated_count[q.number] += 1
+            elif dominates(q, p, directions):
+                dominates_list[q.number].append(p.number)
+                dominated_count[p.number] += 1
+
+        population_per_rank = []
+        while population:
+            non_dominated_population = []
+            i = 0
+            while i < len(population):
+                if dominated_count[population[i].number] == 0:
+                    individual = population[i]
+                    if i == len(population) - 1:
+                        population.pop()
+                    else:
+                        population[i] = population.pop()
+                    non_dominated_population.append(individual)
+                else:
+                    i += 1
+
+            for x in non_dominated_population:
+                for y in dominates_list[x.number]:
+                    dominated_count[y] -= 1
+
+            assert non_dominated_population
+            population_per_rank.append(non_dominated_population)
+
+        return population_per_rank
+
+    def after_trial(
+        self,
+        study: Study,
+        trial: FrozenTrial,
+        state: TrialState,
+        values: Sequence[float] | None,
+    ) -> None:
+        assert state in [TrialState.COMPLETE, TrialState.FAIL, TrialState.PRUNED]
+        if self._constraints_func is not None:
+            _process_constraints_after_trial(self._constraints_func, study, trial, state)
+        self._random_sampler.after_trial(study, trial, state, values)
+
+
+# TODO(Shinichi) Replace with math.comb after support for python3.7 is deprecated.
```

Comment:
The definition of `math.comb` is different from `_multi_choose`.
`math.comb`: `n! / (k! * (n - k)!)`
`_multi_choose`: `(n + k - 1) / (k! * (n-1)!)`

So, we might elaborate this code comment regarding the difference.